### ISSUE TITLE
Support type inference for recursive elements

### ISSUE THREAD

**Issue by joshsh** [CategoricalData/hydra#90]
Hydra's formal data model (Lambda Graph) includes a variant of Hindley-Milner type inference which has been adapted slightly for nominal types (records, unions, and wrappers). However, while λG conceives of graphs as mutually recursive let terms and technically permits any graph which *can* be typed, its type inference does not extend to let terms with actual recursion -- instead, it relies on type annotations to provide the expected type, which are then simply checked for consistency. Ideally, λG would dispense with user-provided type annotations altogether, and simply disallow terms for which type inference fails. That would require a refinement of λG type inference which can in fact deal with recursion. This is to be investigated. @wisnesky has suggested HMX (reference?) constraint-based methods as an alternative to Algorithm W in Hydra.

**Comment by wisnesky** [comment]
Here is the HMX reference http://www.cs.cmu.edu/~aleksey/talks/refinements03/hmx.pdf

**Comment by joshsh** [comment]
Two other links from @wisnesky: [this'n](https://stackoverflow.com/questions/49134312/hindley-milner-type-inference-for-mutually-recursive-functions) and [this'n](https://stackoverflow.com/questions/67254536/how-can-i-infer-types-for-recursive-functions). Apparently, one should just be able to bind the let keys to variables and let unification do its thing.

**Comment by wisnesky** [comment]
Yup, there’s the problem, someone set this thing to “evil”.  https://www.youtube.com/watch?v=Y8dcmLscf3g

**Comment by joshsh** [comment]
All done. Many instances of recursion are now supported without type inference. Some instances involving polymorphism remain hard, and will be the subject of future extensions / refinements.
